[quant] Add backward compatiblity test

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35923 [quant] ConvPackedParams with TorchBind
* **#36639 [quant] Add backward compatiblity test**

Summary:
Add test for serialization to make sure the changes to serialization
are backward compatible

Test Plan:
python test/quantization/test_backward_compatibility.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[quant] Add backward compatiblity test

Summary:
Add test for serialization to make sure the changes to serialization
are backward compatible

Test Plan:
python test/quantization/test_backward_compatibility.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add backward compatiblity test"


Summary:
Add test for serialization to make sure the changes to serialization
are backward compatible

Test Plan:
python test/quantization/test_backward_compatibility.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add backward compatiblity test"


Summary:
Add test for serialization to make sure the changes to serialization
are backward compatible

Test Plan:
python test/quantization/test_backward_compatibility.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add backward compatiblity test"


Summary:
Add test for serialization to make sure the changes to serialization
are backward compatible

Test Plan:
python test/quantization/test_backward_compatibility.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add backward compatiblity test"


Summary:
Add test for serialization to make sure the changes to serialization
are backward compatible

Test Plan:
python test/quantization/test_backward_compatibility.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_backward_compatibility.py:
```diff
@@ -0,0 +1,126 @@
+        self.assertEqual(qmodule_scripted(data), expected, prec=prec)
+        self.assertEqual(qmodule_traced(data), expected, prec=prec)
+
+    @unittest.skipUnless(
```
Can we also run this with qnnpack?

jerryzh168 replied:
not sure, I can allow qnnpack as well and see if it fails in CI

raghuramank100 commented on test/quantization/test_backward_compatibility.py:
```diff
@@ -0,0 +1,126 @@
+        qconv = nniq.ConvReLU3d(3, 3, kernel_size=3, stride=1, padding=0, dilation=1,
+                                groups=1, bias=True, padding_mode="zeros")
+        self._test_op(qconv)
+        # TODO: graph mode quantized conv3d module
```
For graph mode, you would run the same test as above, except that you wouldnt pass a module and check for state-dict compatibility as the model is scripted.  I think the test would be:
```python
       expected = torch.load(expected_file)
       self.assertEqual(qmodule_scripted(data), expected, prec=prec)
       self.assertEqual(qmodule_traced(data), expected, prec=prec)
```

jerryzh168 replied:
yeah, will add this after the serialization is supported for graph mode

raghuramank100 commented on test/quantization/test_backward_compatibility.py:
```diff
@@ -0,0 +1,126 @@
+        " Quantized operations require FBGEMM. FBGEMM is only optimized for CPUs"
+        " with instruction set support avx2 or newer.",
+    )
+    def test_conv3d(self):
```
Can you also add a test for linear and linear_relu?

jerryzh168 replied:
sure

jerryzh168 committed:
Update on "[quant] Add backward compatiblity test"


Summary:
Add test for serialization to make sure the changes to serialization
are backward compatible

Test Plan:
python test/quantization/test_backward_compatibility.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_backward_compatibility.py:
```diff
@@ -0,0 +1,159 @@
+        # saving code : uncomment when adding a new test
+        # and run the new test, e.g.
+        # python test/quantization/test_backward_compatibility.py TestSerialization.test_conv3d
+        data = torch.randn(2, 3).float()
```
This needs to be commented prior to landing the code

raghuramank100 commented on test/quantization/test_backward_compatibility.py:
```diff
@@ -0,0 +1,159 @@
+        expected = torch.load(expected_file)
+        self.assertEqual(qmodule(data), expected, prec=prec)
+        self.assertEqual(qmodule_scripted(data), expected, prec=prec)
+        # self.assertEqual(qmodule_traced(data), expected, prec=prec)
```
Why is this commented?  We should be able to run a traced model:
https://github.com/pytorch/pytorch/blob/3c85f44ce8335753ecf9d33307a013c9d98c6509/torch/testing/_internal/common_quantization.py#L139
Infact, is the restriction referred to in this code still needed? 
https://github.com/pytorch/pytorch/issues/23986 is closed, so this should work.

jerryzh168 replied:
check_trace is removed in the most recent commit

raghuramank100 reviewed (changes requested):
Few more questions, thanks!

jerryzh168 committed:
Update on "[quant] Add backward compatiblity test"


Summary:
Add test for serialization to make sure the changes to serialization
are backward compatible

Test Plan:
python test/quantization/test_backward_compatibility.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

fmassa commented on test/quantization/test_backward_compatibility.py:
```diff
@@ -0,0 +1,159 @@
+
+        # saving code : uncomment when adding a new test
+        # and run the new test, e.g.
+        # python test/quantization/test_backward_compatibility.py TestSerialization.test_conv3d
```
I think the jit expect tests have an `--accept` flag that regenerates the expect files, maybe it would be useful to add this here?

jerryzh168 replied:
Actually, I think adding an extra argument will work for this case

jerryzh168 committed:
Update on "[quant] Add backward compatiblity test"


Summary:
Add test for serialization to make sure the changes to serialization
are backward compatible

Test Plan:
python test/quantization/test_backward_compatibility.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
closed in favior of https://github.com/pytorch/pytorch/pull/36771
